### PR TITLE
Create OIDCTokenCtx method so that we don't reuse context.

### PR DESCRIPTION
**Pull request by stefan-improbable**

### CODE REVIEW

File: gsa/token_src.go [hunk 1]
```diff
@@ -66,10 +66,16 @@ func NewOIDCTokenSource(ctx context.Context, logger *log.Logger, googleServiceAc
 	return reuseTokenSource, func() error { reset(); return nil }, nil
 }
 
+// OIDCToken is the same as OIDCTokenCtx, except it uses context from itself.
+// Deprecated: use OIDCTokenCtx method instead.
+func (s *OIDCTokenSource) OIDCToken() (*oidc.Token, error) {
+	return s.OIDCTokenCtx(s.ctx)
+}
+
 // OIDCToken is used to obtain new OIDC Token (which includes e.g access token and id token).
```
Comment (Bplotka) [inline gsa/token_src.go]: Fix comment -> `OIDCToken` -> `OIDCTokenCtx`

File: login/token_src.go [hunk 2]
```diff
@@ -126,25 +126,31 @@ func (s *OIDCTokenSource) getOIDCConfigWithRedirectURL(redirectURL string) oidc.
 	return oidcConfig
 }
 
+// OIDCToken is the same as OIDCTokenCtx, except it uses context from itself.
```
Comment (Bplotka) [inline login/token_src.go]: Add warning `Do not user per request timeouts in stuct ctx` .. maybe in constructor?

File: login/token_src.go [hunk 3]
```diff
@@ -224,11 +230,11 @@ func (s *OIDCTokenSource) newToken() (*oidc.Token, error) {
 		extra.Set("nonce", nonce)
 	}
 
-	ctx, cancel := context.WithTimeout(s.ctx, 1*time.Minute)
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, 1*time.Minute)
```
Comment (Bplotka) [inline login/token_src.go]: this is interesting one. I guess most of the user context will have stricter timeout than 1 minute here...  Maybe we should not base on user context here?

File: token_src.go [hunk 4]
```diff
@@ -14,10 +14,14 @@ import (
 
 // TokenSource is anything that can return an oidc token and verifier for token verification.
 type TokenSource interface {
```
Comment (Bplotka) [inline token_src.go]: Wonder if we should create `TokenCtxSource` do not break compatibility?

File: login/token_src.go (hunk 3 above)
Comment (stefan-improbable) [reply login/token_src.go]: Don't have much context here :D should I base it on context.Background()?

File: login/token_src.go (hunk 2 above)
Comment (stefan-improbable) [reply login/token_src.go]: done

File: gsa/token_src.go (hunk 1 above)
Comment (stefan-improbable) [reply gsa/token_src.go]: done

File: token_src.go (hunk 4 above)
Comment (stefan-improbable) [reply token_src.go]: done

File: client.go [hunk 5]
```diff
@@ -327,7 +327,8 @@ func (c *Client) ExchangeServiceAccount(ctx context.Context, cfg Config, googleS
 
 // TokenSource returns a TokenSource that returns t until t expires,
 // automatically refreshing it as necessary using the provided context.
-func (c *Client) TokenSource(ctx context.Context, cfg Config, t *Token) TokenSource {
+// Warning: do not use per request timeouts in ctx. Also use OIDCTokenCtx instead.
+func (c *Client) TokenSource(ctx context.Context, cfg Config, t *Token) *ReuseTokenSource {
```
Comment (Bplotka) [inline client.go]: we should not return struct really... Why we cannot return interface?

File: gsa/token_src.go [hunk 6]
```diff
@@ -42,7 +42,7 @@ type OIDCTokenSource struct {
 
 // NewOIDCTokenSource constructs OIDCTokenSource.
 // Only JSON files are supported as ServiceAccount files.
-func NewOIDCTokenSource(ctx context.Context, logger *log.Logger, googleServiceAccountJSON []byte, provider string, cfg OIDCConfig) (src oidc.TokenSource, clearIDToken func() error, err error) {
+func NewOIDCTokenSource(ctx context.Context, logger *log.Logger, googleServiceAccountJSON []byte, provider string, cfg OIDCConfig) (src *oidc.ReuseTokenSource, clearIDToken func() error, err error) {
```
Comment (Bplotka) [inline gsa/token_src.go]: why not leaving it as it was?

File: login/token_src.go [hunk 7]
```diff
@@ -47,7 +47,8 @@ type OIDCTokenSource struct {
 // NewOIDCTokenSource constructs OIDCTokenSource.
 // Note that OIDC configuration can be passed only from cache. This is due the fact that configuration can be stored in cache as well.
 // If the loginServer is nil, login is disabled.
-func NewOIDCTokenSource(ctx context.Context, logger *log.Logger, cfg Config, cache Cache, callbackSrv *CallbackServer) (src oidc.TokenSource, clearIDToken func() error, err error) {
+// Warning: do not use per request timeouts in ctx. Also use OIDCTokenCtx instead.
+func NewOIDCTokenSource(ctx context.Context, logger *log.Logger, cfg Config, cache Cache, callbackSrv *CallbackServer) (src *oidc.ReuseTokenSource, clearIDToken func() error, err error) {
```
Comment (Bplotka) [inline login/token_src.go]: ditto

File: client.go [hunk 8]
```diff
@@ -327,6 +327,7 @@ func (c *Client) ExchangeServiceAccount(ctx context.Context, cfg Config, googleS
 
 // TokenSource returns a TokenSource that returns t until t expires,
 // automatically refreshing it as necessary using the provided context.
+// Warning: do not use per request timeouts in ctx. Also use OIDCTokenCtx instead.
```
Comment (Bplotka) [inline client.go]: kill this - not relevant anymore.

File: client.go [hunk 9]
```diff
@@ -327,6 +327,7 @@ func (c *Client) ExchangeServiceAccount(ctx context.Context, cfg Config, googleS
...
 func (c *Client) TokenSource(ctx context.Context, cfg Config, t *Token) TokenSource {
 	tkr := &TokenRefresher{
 		ctx: ctx,
```
Comment (Bplotka) [inline client.go]: this is wrong isn't? we just should use that. Always pack ctx per method if it needs such. Don't worry about compatibility - we are already breaking it.

File: gsa/token_src.go [hunk 10]
```diff
@@ -66,10 +66,16 @@ func NewOIDCTokenSource(ctx context.Context, logger *log.Logger, googleServiceAc
 	return reuseTokenSource, func() error { reset(); return nil }, nil
 }
 
-// OIDCToken is used to obtain new OIDC Token (which includes e.g access token and id token).
+// OIDCToken is the same as OIDCTokenCtx, except it uses context from itself.
+// Deprecated: use OIDCTokenCtx method instead.
+func (s *OIDCTokenSource) OIDCToken() (*oidc.Token, error) {
```
Comment (Bplotka) [inline gsa/token_src.go]: nuke `OIDCTokenCtx` and use this method with ctx arg

File: gsa/token_src.go [hunk 11]
```diff
@@ -66,10 +66,16 @@ func NewOIDCTokenSource(ctx context.Context, logger *log.Logger, googleServiceAc
...
+	return s.OIDCTokenCtx(s.ctx)
+}
+
+// OIDCTokenCtx is used to obtain new OIDC Token (which includes e.g access token and id token).
 // No refresh token will be returned, because this is token source is only service Accounts and we don't need login for that anyway.
 // No caching is in place. We base for reuse token source to cache valid tokens in memory.
-func (s *OIDCTokenSource) OIDCToken() (*oidc.Token, error) {
+func (s *OIDCTokenSource) OIDCTokenCtx(_ context.Context) (*oidc.Token, error) {
```
Comment (Bplotka) [inline gsa/token_src.go]: please extend `newToken` to pass context and let's pass it correctly

File: login/token_src.go [hunk 12]
```diff
@@ -47,6 +47,7 @@ type OIDCTokenSource struct {
 // NewOIDCTokenSource constructs OIDCTokenSource.
 // Note that OIDC configuration can be passed only from cache. This is due the fact that configuration can be stored in cache as well.
 // If the loginServer is nil, login is disabled.
+// Warning: do not use per request timeouts in ctx. Also use OIDCTokenCtx instead.
```
Comment (Bplotka) [inline login/token_src.go]: kill all these warnings

File: login/token_src.go [hunk 13]
```diff
@@ -47,6 +47,7 @@ type OIDCTokenSource struct {
...
 func NewOIDCTokenSource(ctx context.Context, logger *log.Logger, cfg Config, cache Cache, callbackSrv *CallbackServer) (src oidc.TokenSource, clearIDToken func() error, err error) {
```
Comment (Bplotka) [inline login/token_src.go]: Do NOT pack context into TokenSource struct. We need it in constructor for oidc.NewClient, but NO struct context from now on (:

File: login/token_src.go (hunk 2 above)
Comment (Bplotka) [reply login/token_src.go]: sorry - now with new changes we always use request ones, so we can kill these comments

File: login/token_src.go [hunk 14]
```diff
@@ -126,25 +127,31 @@ func (s *OIDCTokenSource) getOIDCConfigWithRedirectURL(redirectURL string) oidc.
 	return oidcConfig
 }
 
+// OIDCToken is the same as OIDCTokenCtx, except it uses context from itself.
+// Deprecated: use OIDCTokenCtx method instead.
+func (s *OIDCTokenSource) OIDCToken() (*oidc.Token, error) {
```
Comment (Bplotka) [inline login/token_src.go]: nuke `OIDCTokenCtx` and use this method with ctx arg

File: login/token_src.go [hunk 15]
```diff
@@ -126,25 +127,31 @@ func (s *OIDCTokenSource) getOIDCConfigWithRedirectURL(redirectURL string) oidc.
...
+		err = cachedToken.IsValid(ctx, s.Verifier())
 		if err == nil {
 			// Successfully retrieved a non-expired cached token and only if we have ID token as well.
 			return cachedToken, nil
 		}
 		s.logger.Printf("Warn: Cached token is not valid. Cause: %v\n", err)
 		if cachedToken.RefreshToken != "" {
 			// Only if we have refresh token, we can refresh NewIDToken.
-			oidcToken, err := s.refreshToken(cachedToken.RefreshToken)
+			oidcToken, err := s.refreshToken(ctx, cachedToken.RefreshToken)
```
Comment (Bplotka) [inline login/token_src.go]: :+1:

File: login/token_src.go [hunk 16]
```diff
@@ -154,7 +161,7 @@ func (s *OIDCTokenSource) OIDCToken() (*oidc.Token, error) {
 		}
 	}
 	// Our request for access token was denied, either we had no RefreshToken, it was invalid or expired.
-	newToken, err := s.newToken()
+	newToken, err := s.newToken(ctx)
```
Comment (Bplotka) [inline login/token_src.go]: A.W.E.S.O.M.E

File: login/token_src.go [hunk 17]
```diff
@@ -224,11 +231,11 @@ func (s *OIDCTokenSource) newToken() (*oidc.Token, error) {
 		extra.Set("nonce", nonce)
 	}
 
-	ctx, cancel := context.WithTimeout(s.ctx, 1*time.Minute)
+	ctxWithTimeout, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
```
Comment (Bplotka) [inline login/token_src.go]: kill this or use given context instead background ctx

File: token_src.go [hunk 18]
```diff
@@ -60,22 +67,28 @@ func NewReuseTokenSourceWithDebugLogger(ctx context.Context, debugLogger *log.Lo
 	return s, s.reset
 }
 
-// OIDCToken returns the current token if it's still valid, else will
+// OIDCToken is the same as OIDCTokenCtx, except it uses context from itself.
+// Deprecated: use OIDCTokenCtx method instead.
+func (s *ReuseTokenSource) OIDCToken() (*Token, error) {
```
Comment (Bplotka) [inline token_src.go]: nuke `OIDCTokenCtx` and use this method with ctx arg

File: login/token_src.go (hunk 17 above)
Comment (stefan-improbable) [reply login/token_src.go]: I've replaced the background context with ctx from args. Not sure about killing context.WithCancel, you are calling cancel in goroutine after `<-quit` succeeds. Wdyt?

**Comment by Bplotka** [comment]
Also I am not fan of `oidc_testing.Provider` - I would rather fix https://github.com/Bplotka/go-httpt

**Pull request merged**